subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/ExternalResourceResolverDescriptorParseContext.java

```diff
@@ -49,10 +60,25 @@ public ExternalResourceResolverDescriptorParseContext(ComponentResolvers mainRes
     }
 
     public LocallyAvailableExternalResource getMetaDataArtifact(ModuleComponentIdentifier moduleComponentIdentifier, ArtifactType artifactType) {
+        moduleComponentIdentifier = resolveDynamicVersionIfNecessary(moduleComponentIdentifier);
         File resolvedArtifactFile = resolveMetaDataArtifactFile(moduleComponentIdentifier, mainResolvers.getComponentResolver(), mainResolvers.getArtifactResolver(), artifactType);
         return fileResourceRepository.resource(resolvedArtifactFile);
     }
 
+    private ModuleComponentIdentifier resolveDynamicVersionIfNecessary(ModuleComponentIdentifier moduleComponentIdentifier) {
+        ModuleVersionSelector selector = DefaultModuleVersionSelector.newSelector(moduleComponentIdentifier.getGroup(),
+            moduleComponentIdentifier.getModule(),
+            moduleComponentIdentifier.getVersion());
+        DependencyMetadata metadata = new MavenDependencyMetadata(MavenScope.Compile, false, selector, new ArrayList<Artifact>(), new ArrayList<Exclude>());
+        BuildableComponentIdResolveResult idResolveResult = new DefaultBuildableComponentIdResolveResult();
+        mainResolvers.getComponentIdResolver().resolve(metadata, idResolveResult);
+
+        moduleComponentIdentifier = new DefaultModuleComponentIdentifier(moduleComponentIdentifier.getGroup(),
+            moduleComponentIdentifier.getModule(),
+            idResolveResult.getModuleVersionId().getVersion());
```

Comment:
In general, prefer `[Module|Project]Component[Identifier|Selector]` over `ModuleVersion[Identifier|Selector]`. 

So here you can just do: `idResolveResult.getId()` and use the returned `ComponentIdentifier` as input into the next stage of resolution.

A _less good_ alternative is:  `DefaultModuleComponentIdentifier.newId(idResolveResult.getModuleVersionId())`

Reply:
Done. See [here](https://github.com/blindpirate/gradle/blob/464c412b272e0a64bd25d1d9fc0ab71144ec617f/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/ExternalResourceResolverDescriptorParseContext.java#L64)